src/repr/src/relation.rs

```diff
@@ -657,13 +819,285 @@ impl fmt::Display for NotNullViolation {
+    ///
+    /// Note: For building a [`RelationDesc`] see [`RelationDescBuilder::with_column`].
+    #[must_use]
+    pub fn add_column<N, T>(&mut self, name: N, typ: T) -> RelationVersion
+    where
+        N: Into<ColumnName>,
+        T: Into<ColumnType>,
+    {
+        let latest_version = self.latest_version();
+        let new_version = latest_version.bump();
+
+        let next_idx = self.inner.metadata.len();
+        let col_meta = ColumnMetadata {
+            name: name.into(),
+            typ_idx: next_idx,
+            added: new_version,
+            dropped: None,
+        };
+
+        self.inner.typ.column_types.push(typ.into());
+        let prev = self.inner.metadata.insert(ColumnIndex(next_idx), col_meta);
+
+        assert_none!(prev, "column index overlap!");
+        self.validate();
+
+        new_version
+    }
+
+    /// Drops the column `name` from this [`RelationDesc`]. If there are multiple columns with
+    /// `name` drops the left-most one.
+    ///
+    /// # Panics
+    ///
+    /// Panics if a column with `name` does not exist or if the column was already dropped.
+    #[must_use]
+    pub fn drop_column<N>(&mut self, name: N) -> RelationVersion
+    where
+        N: Into<ColumnName>,
+    {
+        let name = name.into();
+        let latest_version = self.latest_version();
+        let new_version = latest_version.bump();
+
+        let col = self
+            .inner
+            .metadata
+            .values_mut()
+            .find(|meta| meta.name == name)
+            .expect("column to exist");
+
+        assert_none!(col.dropped, "column was already dropped");
+        col.dropped = Some(new_version);
+
+        self.validate();
+        new_version
+    }
+
+    /// Returns the [`RelationDesc`] at the latest version.
+    pub fn latest(&self) -> RelationDesc {
+        self.inner.clone()
+    }
+
+    /// Returns this [`RelationDesc`] at the specified version.
+    pub fn at_version(&self, version: RelationVersionSelector) -> RelationDesc {
+        // Get all of the changes from the start, up to whatever version was requested.
+        //
+        // TODO(parkmycar): We should probably panic on unknown verisons?
+        let up_to_version = match version {
+            RelationVersionSelector::Latest => RelationVersion(u64::MAX),
+            RelationVersionSelector::Specific(v) => v,
+        };
+
+        let valid_columns = self.inner.metadata.iter().filter(|(_col_idx, meta)| {
+            let added = meta.added <= up_to_version;
+            let dropped = meta
+                .dropped
+                .map(|dropped_at| up_to_version >= dropped_at)
+                .unwrap_or(false);
+
+            added && !dropped
+        });
+
+        let mut column_types = Vec::new();
+        let mut column_metas = BTreeMap::new();
+
+        // N.B. At this point we need to be careful because col_idx might not
+        // equal typ_idx.
```

Comment:
Why's that? I don't think I understood how col_idx and typ_idx could diverge.

Reply:
Oh because you're not retaining dropped columns in `column_metas` but they are still influencing `col_idx` and not `typ_idx`?

Reply:
> Oh because you're not retaining dropped columns in column_metas but they are still influencing col_idx and not typ_idx?

Exactly! Ideally we would track `ColumnType` in the `BTreeMap<...>` but `RelationType` is used so widely in compute that it's a big refactor. I added an example in the comment and this is exercised in the snapshot tests below, let me know if I should add more!